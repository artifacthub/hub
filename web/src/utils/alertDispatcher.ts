import isUndefined from 'lodash/isUndefined';

import { Alert } from '../types';

const DEFUALT_TIME = 5 * 1000; //5s

export interface AlertUpdatesHandler {
  updateAlertWrapper(alert: Alert | null): void;
}

export class AlertDispatcher {
  private updatesHandler?: AlertUpdatesHandler;
  private visibleAlert: Alert | null = null;
  private timeout?: NodeJS.Timeout;

  public subscribe(updatesHandler: AlertUpdatesHandler) {
    this.updatesHandler = updatesHandler;
  }

  public postAlert(alert: Alert | null) {
    this.visibleAlert = alert;
    this.dismissAlert();
    if (this.updatesHandler) {
      this.updatesHandler.updateAlertWrapper(this.visibleAlert);
    }
  }

  private dismissAlert() {
    this.clearTimeout();
    if (this.visibleAlert && (isUndefined(this.visibleAlert.autoClose) || this.visibleAlert.autoClose)) {
      this.timeout = setTimeout(() => {
        this.postAlert(null);
      }, this.visibleAlert.dismissOn || DEFUALT_TIME);
    }
  }

  private clearTimeout() {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }
  }
}

const alertDispatcher = new AlertDispatcher();
export default alertDispatcher;
