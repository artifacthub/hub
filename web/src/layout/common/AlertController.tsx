import classnames from 'classnames';
import isNull from 'lodash/isNull';
import React, { useState } from 'react';

import { Alert } from '../../types';
import alertDispatcher from '../../utils/alertDispatcher';
import styles from './AlertController.module.css';

const AlertController = () => {
  const [alert, setAlert] = useState<Alert | null>(null);

  alertDispatcher.subscribe({
    updateAlertWrapper: (alert: Alert | null) => setAlert(alert),
  });

  const onClose = () => {
    alertDispatcher.postAlert(null);
  };

  return (
    <div className="position-relative">
      <div
        data-testid="alertController"
        className={classnames(
          `alert alert-dismissible position-fixed`,
          `alert-${!isNull(alert) && alert.type ? alert.type : 'warning'}`,
          styles.alert,
          { [`show ${styles.active}`]: !isNull(alert) },
          { fade: isNull(alert) }
        )}
        role="alert"
      >
        {!isNull(alert) && <>{alert.message}</>}

        <button type="button" className="close" onClick={onClose}>
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
    </div>
  );
};

export default AlertController;
