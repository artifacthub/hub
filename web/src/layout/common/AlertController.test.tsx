import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';

import alertDispatcher from '../../utils/alertDispatcher';
import AlertController from './AlertController';
jest.mock('../../utils/alertDispatcher');

describe('AlertController', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<AlertController />);
    expect(asFragment).toMatchSnapshot();
  });

  it('renders component when alert is null', () => {
    const { getByTestId } = render(<AlertController />);

    const component = getByTestId('alertController');
    expect(component).toBeInTheDocument();
    expect(component).not.toHaveClass('show');
    expect(component).toHaveClass('fade');
    expect(component).toHaveTextContent('×');
  });

  it('closes alert to click close button', () => {
    const { getByText } = render(<AlertController />);

    alertDispatcher.postAlert({
      type: 'warning',
      message: 'alert',
    });

    waitFor(() => {
      const btn = getByText('×');
      fireEvent.click(btn);
      expect(alertDispatcher.postAlert).toHaveBeenCalledTimes(1);
      expect(alertDispatcher.postAlert).toHaveBeenCalledWith(null);
    });
  });

  it('renders warning alert', () => {
    const { getByTestId, getByText } = render(<AlertController />);

    alertDispatcher.postAlert({
      type: 'warning',
      message: 'This is a warning alert',
    });

    waitFor(() => {
      const component = getByTestId('alertController');
      expect(component).toHaveClass('alert-warning');
      expect(component).toHaveClass('show active');

      expect(getByText('This is a warning alert'));
    });
  });

  it('renders success alert', () => {
    const { getByTestId, getByText } = render(<AlertController />);

    alertDispatcher.postAlert({
      type: 'success',
      message: 'This is a success alert',
    });

    waitFor(() => {
      const component = getByTestId('alertController');
      expect(component).toHaveClass('alert-success');
      expect(component).toHaveClass('show active');

      expect(getByText('This is a success alert'));
    });
  });

  it('renders danger alert', () => {
    const { getByTestId, getByText } = render(<AlertController />);

    alertDispatcher.postAlert({
      type: 'danger',
      message: 'This is a danger alert',
    });

    waitFor(() => {
      const component = getByTestId('alertController');
      expect(component).toHaveClass('alert-danger');
      expect(component).toHaveClass('show active');

      expect(getByText('This is a danger alert'));
    });
  });
});
