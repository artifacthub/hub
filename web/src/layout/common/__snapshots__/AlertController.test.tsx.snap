// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertController creates snapshot 1`] = `[Function]`;
